Sources/Extensions/Shared/RandomizableInteger.swift

```diff
@@ -0,0 +1,154 @@
+    ///   - upperBound: The upper bound of the integer range.
+    /// - Returns: An integer between `lowerBound` and `upperBound` (exclusive).
+    static func random(lowerBound: Self, upperBound: Self) -> Self
+}
+
+/// A signed integer that has an unsigned integer counterpart
+public protocol SignedRandomizableInteger: RandomizableInteger, SignedInteger {
+
+    /// The `UnsignedInteger` counterpart with equal bit width
+    associatedtype Unsigned: UnsignedRandomizableInteger
+
+}
+
+/// An unsigned integer that has a signed integer counterpart
+public protocol UnsignedRandomizableInteger: RandomizableInteger, UnsignedInteger {
+
+    /// The `SignedInteger` counterpart with equal bit width
+    associatedtype Signed: SignedRandomizableInteger
+
+}
+
+// MARK: - Random functions
+
+// The randomization functions are adapted from @jstn's solution:
+// https://stackoverflow.com/questions/24007129/how-does-one-generate-a-random-number-in-apples-swift-language/25129039#25129039
+
+public extension UnsignedRandomizableInteger {
+    /// SwifterSwift: Generate a random integer within a given range.
+    ///
+    ///   This version is able to handle `UInt64`,
+    ///   which would not be possible using `BinaryInteger.random(lowerBound:upperBound:)`
+    ///
+    /// - Parameters:
+    ///   - lowerBound: The lower bound of the integer range. Defaults to 0.
+    ///   - upperBound: The upper bound of the integer range.
+    /// - Returns: An integer between `lowerBound` and `upperBound` (exclusive).
+    /// - Complexity:
+    ///   This could theoretically loop forever but each retry has p > 0.5 (worst case, usually far better).
+    public static func random(lowerBound: Self = 0, upperBound: Self) -> Self {
+        // Static asserts are not yet available in Swift
+        assert(Self.bitWidth == Signed.bitWidth)
+        assert(lowerBound < upperBound)
+
+        var mod: Self
+        let diff = upperBound - lowerBound
+        var rand: Self = arc4random()
+
+        if diff > Self(Signed.max) {
+            mod = 1 + ~diff
+        } else {
+            mod = ((max - (diff * 2)) + 1) % diff
+        }
+
+        while rand < mod {
+            rand = arc4random()
+        }
+
+        return (rand % diff) + lowerBound
+    }
+}
+
+public extension SignedRandomizableInteger {
+    /// SwifterSwift: Generate a random integer within a given range.
+    ///
+    ///   This version is able to handle `Int64`, as well as finding a random number in the range `Self.min..<Self.max`,
+    ///   which would not be possible using `BinaryInteger.random(lowerBound:upperBound:)` due to overflow
+    ///
+    /// - Parameters:
+    ///   - lowerBound: The lower bound of the integer range. Defaults to 0.
+    ///   - upperBound: The upper bound of the integer range.
+    /// - Returns: An integer between `lowerBound` and `upperBound` (exclusive).
+    public static func random(lowerBound: Self = 0, upperBound: Self) -> Self {
```

Comment:
Would it make sense to give the `upperBound` parameter a default value of `Self.max`? That static property is defined in by `FixedWidthInteger` which `RandomizableInteger ` and requires conformance to.

Reply:
For the general use case I doubt it would make sense. Often when you want a random number it would be in the context of something like:
`let diceRoll = Int.random(upperBound: 6)`
or
`let randomElement = myArray[Int.random(upperBound: myArray.count)]` (although my next step is to get better random elements from collections 😉)

That opens the question of whether or not the function signature should be:
`random(_ lowerBound: Self = 0, _ upperBound: Self) -> Self`
That's a matter of taste, so I'd appreciate anybody else's input?